### PR TITLE
Tweak ExecutorTestHelper to test pipelines

### PR DESCRIPTION
**Pull request by markuspf**
Proposed changes to ExecutorTestHelper to allow testing sequences of executors.

The main purpose of this is to be able to run integration tests of `SubqueryStartExecutor`/`SubqueryEndExecutor` without having to copy or rewrite `ExecutorTestHelper`

### CODE REVIEW

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8654/

File: tests/Aql/ExecutorTestHelper.h [hunk 1]
```diff
@@ -222,7 +264,33 @@ struct ExecutorTestHelper {
     _expectedStats = stats;
     _testStats = true;
     return *this;
-  };
+  }
+
+  template <typename E>
+  auto setExecBlock(typename E::Infos infos) -> ExecutorTestHelper& {
+    // TODO: this unique_ptr goes out of scope and is free'd, but I think
```
Comment (mchacki) [inline tests/Aql/ExecutorTestHelper.h]: I think you can remove this todo.
The unique_ptr is handled in the _execNodes.
Then you just have a reference to the unique_pointer, which is not deleted here, but on destruction of _execNodes.

Important is that. testeeNode.get() still yields a valid pointer, which in my opinion it does here.

File: tests/Aql/ExecutorTestHelper.h [hunk 2]
```diff
@@ -222,7 +264,33 @@ struct ExecutorTestHelper {
...
+    auto& testeeNode = _execNodes.emplace_back(std::move(
+        std::make_unique<SingletonNode>(_query.plan(), _execNodes.size())));
+    _testee = std::make_unique<ExecutionBlockImpl<E>>(_query.engine(),
+                                                      testeeNode.get(), std::move(infos));
+    return *this;
+  }
+
+  template <typename E>
+  auto createExecBlock(typename E::Infos infos) -> ExecBlock {
+    // TODO: this unique_ptr goes out of scope and is free'd, but I think
```
Comment (mchacki) [inline tests/Aql/ExecutorTestHelper.h]: Same as above.

File: tests/Aql/ExecutorTestHelper.h [hunk 3]
```diff
@@ -284,6 +385,29 @@ struct ExecutorTestHelper {
   };
 
  private:
+  void testOutputBlock(SharedAqlItemBlockPtr const& outputBlock,
```
Comment (mchacki) [inline tests/Aql/ExecutorTestHelper.h]: I think this is not required anymore?
If you modified something in this part, please make sure to move it into the validateBlock asserter class above

File: tests/Aql/ExecutorTestHelper.h [hunk 4]
```diff
@@ -244,19 +312,52 @@ struct ExecutorTestHelper {
...
 
-  auto run(typename E::Infos infos) -> void {
+  auto run() -> void {
     ResourceMonitor monitor;
     AqlItemBlockManager itemBlockManager(&monitor, SerializationFormat::SHADOWROWS);
 
     auto inputBlock = generateInputRanges(itemBlockManager);
 
-    auto testeeNode = std::make_unique<SingletonNode>(_query.plan(), 1);
+      _testee->addDependency(inputBlock.get());
```
Comment (mchacki) [inline tests/Aql/ExecutorTestHelper.h]: Can you unify run and runPipeline?

e.g.:
run() -> 
ASSERT pipeline is empty.
Insert 1 node in _pipeline.
call runPipeline()

I think they should be identical then

File: tests/Aql/ExecutorTestHelper.h (hunk 1 above)
Comment (markuspf) [reply tests/Aql/ExecutorTestHelper.h]: Yes, the comment was written before _execNodes existed.

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8656/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8657/

**Pull request merged**